### PR TITLE
Use new nutrients taxonomy to replace hardcoded values from Food.pm - Ready for review

### PR DESCRIPTION
**Pull request by stephanegigandet**
In a nutshell, the goal is to remove the hardcoded translations and properties for nutrients that are currently in Food.pm, and use a nutrients taxonomy instead.

- Moved @aleene's nutrients taxonomy to nutrients_old.txt,
- Created a script to generate a new taxonomy source file nutrients.txt, from the hardcoded values in Food.pm (that changed since the old nutrients.txt taxonomy was created) + the old nutrients taxonomy in order to get all the synonyms from wikidata.
- Replaced all the uses of %Nutriments (the hardcoded values in Food.pm) with the new nutrients taxonomy
- Had to move some code around between Food.pm, Tags.pm and the new PackagerCodes.pm module
- Simplified a bit the template for the product edit form
- Removed %Nutriments from Food.pm

Note: the canonical entry for nutrients in the new taxonomy use the zz: language prefix. This is because we can't change the ids of nutrients from the ones we have today in the nutriments hash of the products, even though they do not correspond exactly to their English names. e.g. some are shortened ("fruits-vegetable-nuts" instead of "fruits, vegetables, nuts, and rapeseed, walnut and olive oil").

### CODE REVIEW

**Review by alexgarel — COMMENTED** [review]
I have put some comment, I would like to test tomorrow before approving.

File: lib/ProductOpener/Food.pm [hunk 1]
```diff
@@ -242,13 +229,13 @@ sub assign_nid_modifier_value_and_unit($$$$$) {
 	$product_ref->{nutriments}{$nid . "_unit"} = $unit;
 	$product_ref->{nutriments}{$nid . "_value"} = $value;
 	
-	if (((uc($unit) eq 'IU') or (uc($unit) eq 'UI')) and (exists $Nutriments{$nid}) and ($Nutriments{$nid}{iu} > 0)) {
-		$value = $value * $Nutriments{$nid}{iu} ;
-		$unit = $Nutriments{$nid}{unit};
+	if (((uc($unit) eq 'IU') or (uc($unit) eq 'UI')) and (defined get_property("nutrients", "zz:$nid", "iu_value:en"))) {
+		$value = $value * get_property("nutrients", "zz:$nid", "iu_value:en") ;
+		$unit = get_property("nutrients", "zz:$nid", "iu_value") || 'g';
```
[the remark this answers is not in the document]
Comment (stephanegigandet) [reply lib/ProductOpener/Food.pm:234]: Right it should be iu_value:en

We could use // indeed, although iu_value, if defined, is never 0.

File: cgi/product_multilingual.pl [hunk 2]
```diff
@@ -1254,30 +1254,22 @@ ($$$$$)
...
 		$nutriment_ref->{prefix} = $prefix;
 
-		my $unit = 'g';
-		if ((exists $Nutriments{$nid}) and (exists $Nutriments{$nid}{"unit_$cc"})) {
-			$unit = $Nutriments{$nid}{"unit_$cc"};
+		my $unit = "g";
+
+		if (exists_taxonomy_tag("nutrients", "zz:$nid")) {
+			$nutriment_ref->{name} = display_taxonomy_tag($lc, "nutrients", "zz:$nid");
+			$unit = get_property("nutrients", "zz:$nid", "unit_$cc:en") // get_property("nutrients", "zz:$nid", "unit:en") // 'g';
```
[the remark this answers is not in the document]
Comment (stephanegigandet) [reply cgi/product_multilingual.pl:1265]: $cc is a global variable, set to the country code. e.g. on fr.openfoodfacts.org, it is "fr". We have a similar $lc variable for the language code. I'll add a comment.

File: lib/ProductOpener/Food.pm (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephanegigandet) [reply lib/ProductOpener/Food.pm:234]: In fact we check the property is defined just before, so we don't need the || or //, I'll remove it.

File: lib/ProductOpener/Food.pm [hunk 3]
```diff
@@ -1079,2926 +1049,31 @@ sub mmoll_to_unit {
...
+	my $tagid = "zz:" . $nid;
+	my $error = 0;
 
+	if (not exists_taxonomy_tag("nutrients", $tagid)) {
+		print STDERR "Nutrient $nid is defined in \%nutriments_table but does not exist in the nutrients taxonomy.\n";
+		$error++;
+	}
+	if ($error) {
+		die("Some nutrients are defined in \%nutriments_table but do not exist in the nutrients taxonomy.");
+	}
```
[the remark this answers is not in the document]
Comment (stephanegigandet) [reply lib/ProductOpener/Food.pm:1075]: I'll move it to food.t

File: lib/ProductOpener/Food.pm [hunk 4]
```diff
@@ -4027,80 +1102,57 @@ foreach my $region (keys %nutriments_tables) {
...
+			my $tag = sprintf($fmt, $nutrient_name, $level_quantity);
+			my $tagid = get_string_id_for_lang($lc, $tag);
+			$canon_tags{$lc}{nutrient_levels}{$tagid} = $tag;
+			# print "nutrient_levels : lc: $lc - tagid: $tagid - tag: $tag\n";
+		}
+	}
+}
 
-	foreach my $lc (sort keys %{$Nutriments{$nid}}) {
+$log->debug("Nutrient levels initialized") if $log->is_debug();
```
[the remark this answers is not in the document]
Comment (stephanegigandet) [reply lib/ProductOpener/Food.pm:1133]: We could, but that means we also need to call that function in all the scripts in /scripts that use Food.pm. I

File: lib/ProductOpener/Food.pm [hunk 5]
```diff
@@ -5199,7 +2258,7 @@ sub compute_unknown_nutrients($) {
 
 		next if $nid =~ /_/;
 
-		if ((not exists $Nutriments{$nid}) and (defined $product_ref->{nutriments}{$nid . "_label"})) {
+		if ((not exists_taxonomy_tag("nutrients", "zz:" . $nid)) and (defined $product_ref->{nutriments}{$nid . "_label"})) {
```
[the remark this answers is not in the document]
Comment (stephanegigandet) [reply lib/ProductOpener/Food.pm:2261]: ok, will do

File: lib/ProductOpener/Food.pm (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexgarel) [reply lib/ProductOpener/Food.pm:1105]: Or call it at end of file at least, to be able to move it later on ?

File: lib/ProductOpener/Food.pm (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexgarel) [reply lib/ProductOpener/Food.pm]: cool.

File: cgi/product_multilingual.pl (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexgarel) [reply cgi/product_multilingual.pl:1266]: yes ok cc for country code

**Review by alexgarel — APPROVED** [review]
Tested ok.

**Pull request merged**